onmt/translate/penalties.py

```diff
@@ -1,80 +1,57 @@
 from __future__ import division
-import torch
 
 
-class PenaltyBuilder(object):
+def coverage_penalty(name):
+    coverages = {"wu": coverage_wu, "summary": coverage_summary}
+    return coverages.get(name, coverage_none)
+
+
+def length_penalty(name):
+    lengths = {"wu": length_wu, "average": length_average}
+    return lengths.get(name, length_none)
+
+
+def coverage_wu(scores, cov, beta=0.):
+    """
+    NMT coverage re-ranking score from
+    "Google's Neural Machine Translation System" :cite:`wu2016google`.
+    """
+    return 0. if beta == 0. else beta * -cov.clamp(max=1.0).log().sum(1)
+
+
+def coverage_summary(scores, cov, beta=0.):
+    """
+    Our summary penalty.
     """
-    Returns the Length and Coverage Penalty function for Beam Search.
+    return 0. if beta == 0. else beta * \
+        (-cov.clamp(min=1.0).sum(1) - cov.size(1))
```

Comment:
This triggers an error, clamp does not exist.

Reply:
[Clamp](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.clamp) [exists](https://pytorch.org/docs/stable/torch.html#torch.clamp). What is the value of `cov`?

Reply:
5.0 it's a flaot in the opts.py

Reply:
Looking at the history of the penalties code, that must be a recent change.

Reply:
I think there is a confusion here - `beta` is the hyperparameter that is a float. `cov` is the sum of attentions from all previous decoding steps. 
The summary penalty is from my recent EMNLP paper and penalizes repeated attention, wu penalty is from the google nmt paper and penalizes not attenting to words. 
@vince62s What is the exact error you are getting?

Reply:
my confusion.
the issue is not here but in beam.py below.